### PR TITLE
drivers: adc: stm32h7: remove oversampling workaround

### PR DESCRIPTION
**Pull request by heinwessels**
Removes a workaround that was required while there was a bug in upstream STM32 HAL drivers where the oversampling was not set correctly in all cases. This has since been fixed, and also added to the Zephyr STM32 HAL.

The upstream issue:
https://github.com/STMicroelectronics/STM32CubeH7/issues/177

The fix was added to stm32_hal here:
https://github.com/zephyrproject-rtos/hal_stm32/commit/55648728d4ca9d17595d6adc2db447438c48bfef

The fixed code:
https://github.com/zephyrproject-rtos/hal_stm32/blob/5c79d0c7c4f7094e19959f0bc542ebc35db44002/stm32cube/stm32h7xx/drivers/include/stm32h7xx_ll_adc.h#L6389-L6405

### CODE REVIEW

**Review by gautierg-st — CHANGES_REQUESTED** [review]

File: drivers/adc/adc_stm32.c [hunk 1]
```diff
@@ -627,21 +625,13 @@ static void adc_stm32_oversampling(ADC_TypeDef *adc, uint8_t ratio, uint32_t shi
...
-		MODIFY_REG(adc->CFGR2, (ADC_CFGR2_OVSS | ADC3_CFGR2_OVSR),
-			(shift | stm32_adc_ratio_table[ratio]));
+	if (adc != ADC3) {
+		/* the LL function expects a value from 1 to 1024 */
+		adc_stm32_oversampling_ratioshift(adc, 1 << ratio, shift);
 	} else {
-		MODIFY_REG(adc->CFGR2, (ADC_CFGR2_OVSS | ADC_CFGR2_OVSR),
-			(shift | (((1UL << ratio) - 1) << ADC_CFGR2_OVSR_Pos)));
+		/* the LL function expects a value LL_ADC_OVS_RATIO_x */
+		adc_stm32_oversampling_ratioshift(adc, stm32_adc_ratio_table[ratio], shift);
```
Comment (gautierg-st) [inline drivers/adc/adc_stm32.c:633]: Sorry, I haven't seen that in my first review, but you also need to check which version of ADC3 you have. Only the ADC3 of STM32H72x/H73x is impacted by this, for the rest of the H7 family ADC3 behaves like ADC1/2.
You need to have a `#ifdef ADC_VER_V5_V90` and proceed like in the `LL_ADC_ConfigOverSamplingRatioShift` in stm32h7xx_ll_adc.h.

**Review by gautierg-st — CHANGES_REQUESTED** [review]

File: drivers/adc/adc_stm32.c (hunk 1 above)
Comment (heinwessels) [reply drivers/adc/adc_stm32.c:633]: Ah, you're right. I wrongly assumed the fix would allow the call the be more generic. Should've caught that, my bad.

**Review by gautierg-st — DISMISSED** [review]
LGTM

**Review by erwango — DISMISSED** [review]
Please fix spelling issue reported.
Otherwsie LGTM

**Comment by heinwessels** [comment]
> Please fix spelling issue reported. Otherwsie LGTM

Fixed the spelling mistake.

**Review by gautierg-st — APPROVED** [review]

**Review by erwango — APPROVED** [review]

**Pull request merged**